torch/csrc/finalizer.h

```diff
@@ -0,0 +1,23 @@
+#pragma once
+
+#include <Python.h>
+#include <TH/THStorage.hpp>
+#include <torch/csrc/utils/object_ptr.h>
+
+namespace torch {
+
+struct PyObjectFinalizer : public THFinalizer {
+  THPObjectPtr pyobj_;
```

Comment:
This looks like it will call decrement the Python reference count in the destructor, which might not be holding the GIL?

Reply:
You're right, this is a bug.

Reply:
Isn't this a hella big hazard with THPObjectPtr? Shouldn't we make its destructor automatically take out the GIL?

Reply:
Yeah, it's a hazard, but most use case are on the stack (i.e. implicitly thread-local), so I'm not sure it's worth the perf hit.